planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,210 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"container/list"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/kvcache"
+)
+
+// CacheEntry wraps Key and Value. It's the value of list.Element.
+type CacheEntry struct {
+	PlanKey   kvcache.Key
+	PlanValue kvcache.Value
+}
+
+// LRUPlanCache is a dedicated least recently used cache, JUST use for plan cache.
+type LRUPlanCache struct {
+	capacity uint
+	size     uint
+	// buckets replace the map in general LRU
+	buckets map[hack.MutableString]map[*list.Element]struct{}
+	cache   *list.List
+	// lock make cache thread safe
+	lock sync.Mutex
+
+	// pickFromBucket get one element from bucket according to the incoming function.The LRUPlanCache can not work if this is nil
+	pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool)
+	// onEvict will be called if any eviction happened
+	onEvict func(kvcache.Key, kvcache.Value)
+}
+
+// NewLRUPlanCache creates a PCLRUCache object, whose capacity is "capacity".
+// NOTE: "capacity" should be a positive value.
+func NewLRUPlanCache(capacity uint, pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool),
+	onEvict func(kvcache.Key, kvcache.Value)) *LRUPlanCache {
+	if capacity < 1 {
+		panic("capacity of LRU Cache should be at least 1.")
+	}
+	return &LRUPlanCache{
+		capacity:       capacity,
+		size:           0,
+		buckets:        make(map[hack.MutableString]map[*list.Element]struct{}),
+		cache:          list.New(),
+		pickFromBucket: pickFromBucket,
+		onEvict:        onEvict,
+	}
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *LRUPlanCache) Get(key kvcache.Key, paramTypes []*types.FieldType) (value kvcache.Value, ok bool) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	bucket, bucketExist := l.buckets[hack.String(key.Hash())]
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			l.cache.MoveToFront(element)
+			return element.Value.(*CacheEntry).PlanValue, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *LRUPlanCache) Put(key kvcache.Key, value kvcache.Value, paramTypes []*types.FieldType) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := hack.String(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	// bucket exist
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			element.Value.(*CacheEntry).PlanValue = value
+			l.cache.MoveToFront(element)
+			return
+		}
+	} else {
+		l.buckets[hash] = make(map[*list.Element]struct{})
```

Comment:
How about making the map with an initialized size:
```suggestion
		l.buckets[hash] = make(map[*list.Element]struct{}, 1)
```